options: improve watch later menu

yatinlala opened the pull request:
The History menu (CTRL-P -> History) displays the titles of previously played files, but the Watch later menu displays a list of filenames.

This PR adds `--write-title-in-watch-later-config`. With the option set, `save-position-on-quit` inserts the playing file's title in a comment below the filename in the file's watch_later entry. The Watch later picker then displays these titles instead of filenames when they are available. 


## Examples
<ul>
<li><code>mpv --write-filename-in-watch-later-config</code>:</li>
<pre># https://example.com/?v=foobar<br>start=35<br>gamma=1</pre>
<li><code>mpv --write-filename-in-watch-later-config --write-title-in-watch-later-config</code>:</li>
<pre># https://example.com/?v=foobar<br># FooBar (HD Video)<br>start=35<br>gamma=1</pre>
<li><code>mpv --write-title-in-watch-later-config</code>:</li>
<pre>start=35<br>gamma=1</pre>
</ul>


## Notes:
- `--write-title-in-watch-later-config` silently no-ops if  `--write-filename-in-watch-later config` is not set.  The Watch later menu is disabled without `--write-filename-in-watch-later-config` anyway, so this seems like a reasonable behaviour to me. If it is preferred that `--write-title...` automatically enables `--write-filename...` or emits an error, let me know.

- These changes are backwards compatible with older versions of MPV, since, forcing `--write-filename...` to be set ensures that a title comment will never be inserted without a filename comment above it. Title comments will never be misinterpreted as paths to a file by older versions.

- Titles are displayed instead of filenames in the Watch later menu if `--write-title-in-watch-later-config` is set, otherwise existing behaviour is preserved. Perhaps this should be separated out into its own option, applying to both the History and Watch later menus?

- I acknowledge this is a hacky solution. If breaking backwards compatibility is an option, it might be nice to replace the watch_later system with a structured data format in line with `watch_history.jsonl`.

yatinlala committed:
player/configfiles: write title to watch_later file

write_title() writes the media title to the media's watch_later file
as a comment if 'write-filename-in-watch-later-config' is set. Prior to
this change, setting the option would only write a commented filename at
the top of the file. Now the media title is commented below the
filename, with the prefix "title: "

Before:
 # https://example.com/watch?v=foobar
 start=35
 gamma=1

After:
 # https://example.com/watch?v=foobar
 # title: Foo Bar (HD Video)
 start=35
 gamma=1

yatinlala committed:
select.lua: display commented titles in watch_later files

If --write-filename-in-watch-later-config is set and a title is written
in the watch later file, display it in the Watch later menu.

guidocella commented:
Thanks for the patch, I also thought this would be useful for web URLs and this was requested on IRC.

Some thoughts:
- Do we need a new option instead of just always saving the title with `--write-filename-in-watch-later-config`, like saving history always saves the title? Is there a reason to save the filename but not the title?
- Can we call `format_history_entry()` so the format is consistent with history? Which also allows customizing the date format

> * I acknowledge this is a hacky solution. If breaking backwards compatibility is an option, it might be nice to replace the watch_later system with a structured data format in line with `watch_history.jsonl`.

I chose jsonl for history because it is only appended too, it is slow if you rewrite the whole file. That is why the single-file shader cache was replaced with multiple files. In fact doing the opposite was requested in
https://github.com/mpv-player/mpv/issues/16429#issuecomment-2954459804

guidocella commented:
Also maybe we should prefix `title=` so we can add other data in the future. I know @Earnestly saves the playlist title with a custom script for selecting youtube playlists, maybe he has suggestions?

guidocella commented:
Ok I guess the `title=` prefix is not needed since the playlist title would be stored separately in the redirect entry. I think this PR can just store the video title, and a future PR can resurrect #13476 and also store the playlist title after the playlist path. Then redirect entries with a title can be added to the menu.

Also unlike with json this all breaks with paths and titles containing newlines, but they are not common anyway.

Earnestly commented:
> Also unlike with json this all breaks with paths and titles containing newlines, but they are not common anyway.

Newlines aren't the problem with json, however json does require valid UTF-8 which is not required by unix filenames.

yatinlala committed:
DOCS/man/options: write-filename-in-watch-later-config also writes title

yatinlala commented:
Thank you for all the feedback!

> - Do we need a new option instead of just always saving the title with --write-filename-in-watch-later-config, like saving history always saves the title? Is there a reason to save the filename but not the title?

You're right! Storing filename without title certainly seems pointless. I could imagine someone wanting to be able to choose between displaying titles and filenames in the menu, but it is a separate thing. I've rolled the change into `--write-filename-...` if that is alright with you.

> - Can we call format_history_entry() so the format is consistent with history? Which also allows customizing the date format

Done.

> - Ok I guess the title= prefix is not needed since the playlist title would be stored separately in the redirect entry.

I haven't added a prefix as per your discussion. It seems to me that a prefix would make it a bit nicer to add more metadata down the line, but I'm uneducated about how playlists work and what additional metadata you might want to store. Let me know if you have a contrary preference.

guidocella commented:
> I haven't added a prefix as per your discussion. It seems to me that a prefix would make it a bit nicer to add more metadata down the line, but I'm uneducated about how playlists work and what additional metadata you might want to store. Let me know if you have a contrary preference.

Dunno I can't think of other metadata we may need. If you or someone else thinks a prefix is better go ahead and add it.

yatinlala commented:
One last change — apologies @guidocella — 6a0454c753a1046b06df9cfda3e5406fe2717808 makes select.lua check all lines after url line for `# title:`. This way we can have unordered metadata in the future that relies on the prefix to identify what it is, instead of hardcoding line numbers.

guidocella commented:
Ok I don't think that needs to be done preemptively but whatever you prefer. Do you have other metadata to store in mind?

yatinlala commented:
Nothing specific in mind. Thumbnail paths or arbitrary user metadata via the Lua API could be cool, but is perhaps unneeded. I'm not familiar enough with the project to know what is possible already and what makes sense to add. 🙂 

I do think that it beats the point of a prefix to assume specific pieces of metadata are coming on specific lines (beyond perhaps better readability).  If hardcoding the line number is the way to go, perhaps I should get rid of the prefix all together.

guidocella committed:
player/configfiles: write playlist titles in watch later files

If ytdl_hook extracted a playlist title, write it in the watch later
redirect entry. This will be shown in the watch later menu.

guidocella committed:
select.lua: show playlists with a title in watch later menu

Add redirect entries with a title to the watch later menu, e.g. youtube
playlists.

Checking for the title avoids showing redirect entries of local
directories, which would drastically increase the number of entries,
when you can already play adjacent local files with
--autocreate-playlist.

yatinlala commented:
Sorry for abandoning the branch 🙏

@guidocella I've added in your patch and updated your email in the commits. Please let me know if I'm missing any changes that need to be made.

I'd love to get this PR merged if its useful or closed otherwise.

guidocella commented:
You can squash the duplicate commits but otherwise looks good.

yatinlala commented:
I believe I have addressed your concerns and rebased on main. I've also tested w/ individual playlists and videos and everything works.

guidocella commented:
What do you think about removing the last 2 commits to make this more likely to be reviewed? The playlist saving also has the issues that after finishing a video you have to wait for the next one to load before quitting to make the playlist title available again, and resolving the playlist entries every time is slow. I am now saving the resolved playlist URLs in a local .m3u file instead with:

```lua
mp.add_key_binding('Alt+s', 'save-playlist', function ()
    local m3u = '#EXTM3U\n'

    for _, entry in ipairs(mp.get_property_native('playlist')) do
        m3u = m3u .. '#EXTINF:-1,' .. (entry.title or '') .. '\n' .. entry.filename .. '\n'
    end

    local file = io.open(os.getenv('HOME') .. '/' .. mp.get_property('metadata/by-key/ytdl_playlist_title', 'playlist') .. '.m3u', 'w')
    if file then
        file:write(m3u)
        file:close()
        mp.osd_message('Playlist saved')
    else
        mp.osd_message('Failed to save the playlist')
    end
end)
```

yatinlala commented:
Makes sense to me -- I've dropped the last two commits and quickly tested on my laptop.
